File: Algorithm/java/class06/GraphGenerator.java
package class06;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class GraphGenerator {
    
    // matrix 所有的边
    //  N * 3矩阵，N行3列
    // [from, to, weight] 
    /**
     * @param matrix
     * @return
     */
    public static Graph createGraph(Integer[][] matrix){
        Graph graph = new Graph();
        for(int i = 0; i < matrix.length; i++){
            Integer from    = matrix[i][0];
            Integer to      = matrix[i][1];
            Integer weight  = matrix[i][2];
            if(!graph.nodes.containsKey(from)){
                graph.nodes.put(from, new Node(from));
            }
            if(!graph.nodes.containsKey(to)){
                graph.nodes.put(to, new Node(to));
            }
            Node fromNode = graph.nodes.get(from);
            Node toNode   = graph.nodes.get(to);
            Edge newEdge = new Edge(weight, fromNode, toNode);
            fromNode.nexts.add(toNode);
            fromNode.out++;
            toNode.in++;
            fromNode.edges.add(newEdge);
            graph.edges.add(newEdge);
        }

        return graph;
    }

    // 从文件中提取数据并创建一个矩阵
    public static Integer[][] createMatrix(String filePath){
        int[] dimension = getFileRowAndColNumer(filePath);
        int row = dimension[0], col = dimension[1];
        Integer[][] matrix = new Integer[row][col];
        try(BufferedReader reader = new BufferedReader(new FileReader(filePath))){
            String line;
            int i = 0, j = 0;
            while ((line = reader.readLine()) != null) {
                String[] values = line.split(" ");
                while (j < col && i < row) {
                    matrix[i][j] = Integer.parseInt(values[j]);
                    j++;
                }
                i++;
                j = 0;
            }
        } catch (IOException e){
            e.printStackTrace();
        }
        return matrix;
    }

    public static void printMatrix(Integer[][] matrix){
        for(int i = 0; i < matrix.length; i++){
            for(int j = 0; j < matrix[0].length; j++){
                if(matrix[i][j] == null){
                    break;
                }
                System.out.print(matrix[i][j] + " ");
                
            }
            System.err.println();
            if(matrix[i][0] == null){
                break;
            }
        }
    
    }

    // 返回一个长度为2的数组[row, col]，包含矩阵的行数和列数
    public static int[] getFileRowAndColNumer(String filePath){
        int row = 0, col = 0;
        try(BufferedReader reader = new BufferedReader(new FileReader(filePath))){
            String line = reader.readLine();
            row++;
            String[] values = line.split(" ");
            col = values.length;
            while ((line = reader.readLine()) != null) {
                row++;
            }
            
        } catch (IOException e){
            e.printStackTrace();
        }
        int[] res = {row, col};
        return res;
    }

    public static void main(String[] args) {
    
        String filePath = "/home/ubuntu/htl/Algorithm/java/class06/matrix.txt";
        Integer[][] matrix = createMatrix(filePath);
        printMatrix(matrix);

        int[] res = getFileRowAndColNumer(filePath);
        System.err.print("row = " + res[0] + " col = " + res[1] + "\n");
        
    }
}

File: Algorithm/python/class06/GraphGenerator.py
"""构造一个图"""
import numpy as np

import Graph, Node, Edge

class Queue:
    def __init__(self):
        self.queue = []
    def add(self, e):
        self.queue.append(e)
    def pop(self):
        if len(self.queue) == 0:
            return None
        res = self.queue[0]
        self.queue = self.queue[1:]
        return res
    def isEmpty(self):
        if len(self.queue) == 0:
            return True
        else:
            return False

class Stack:
    def __init__(self):
        self.stack = []
    def push(self, e):
        self.stack.append(e)
    def pop(self):
        if len(self.stack) == 0:
            return None
        res = self.stack.pop()
        return res
    def isEmpty(self):
        if len(self.stack) == 0:
            return True
        else:
            return False
        

class GraphGenerator:
    def createGraph(self, matrix):
        self.graph = Graph.Graph()
        for i in range(len(matrix)):
            fromNum = matrix[i][0]
            toNum   = matrix[i][1]
            weight   = matrix[i][2]
            if not fromNum in self.graph.nodes:
                self.graph.nodes[fromNum] = Node.Node(fromNum)
            if not toNum in self.graph.nodes:
                self.graph.nodes[toNum] = Node.Node(toNum)
            fromNode = self.graph.nodes[fromNum]
            toNode   = self.graph.nodes[toNum]
            newEdge  = Edge.Edge(fromNode, toNode, weight)
            fromNode.nexts.append(toNode)
            fromNode.outNum += 1
            toNode.inNum += 1
            fromNode.edges.append(newEdge)
            self.graph.edges.add(newEdge)
        return self.graph

    # 从文件中提取数据来创建一个矩阵
    def createMatrix(self, filePath):
        row, col = self.getFileRowAndColNumber(filePath)
        i, j = 0, 0
        matrix = np.zeros((row, col), dtype=int)
        with open(filePath, 'r') as file:
            lines = file.readlines()
            for line in  lines:
                line = line.strip()
                values = line.split()
                while j < col:
                    matrix[i][j] = values[j]
                    j = j + 1
                i = i + 1
                j = 0
        return matrix

    def printMatrix(self, matrix):
        row, col = len(matrix), len(matrix[0])
        for i in range(row):
            for j in range(col):
                print(f"{matrix[i][j]} ", end="")
            print()
    # 得到txt文件的行数和列数（其中列数不包括空格）
    def getFileRowAndColNumber(self, filePath):
        with open(filePath, 'r') as file:
            lines = file.readlines()
            row = len(lines)
            line = lines[0]
            values = line.split(" ")
            col = len(values)
            return row, col


    # 实现图的宽度优先遍历（广度优先搜索）
    def bfs(self, node):
        if(node == None):
            return
        queue = Queue()
        setnode = set()
        queue.add(node)
        setnode.add(node)
        while not queue.isEmpty():
            cur = queue.pop()
            print(f"{cur.value} ", end="")
            for next in cur.nexts:
                if not next in setnode:
                    setnode.add(next)
                    queue.add(next)
        print()


    def dfs(self, node):
        if node == None:
            return
        stack = Stack()
        nodeset = set()
        stack.push(node)
        nodeset.add(node)
        print(f"{node.value} ", end="")
        while not stack.isEmpty():
            cur = stack.pop()
            for next in cur.nexts:
                if not next in nodeset:
                    stack.push(cur)
                    stack.push(next)
                    nodeset.add(next)
                    print(f"{next.value} ", end="")
                    break
        print()
    def sortedTopolgy(self, graph):
        inMap = {}
        zeroInQueue = Queue()
        for node in self.graph.nodes.values():
            inMap[node] = node.inNum
            if(node.inNum == 0):
                zeroInQueue.add(node)
        result = []
        while not zeroInQueue.isEmpty():
            cur = zeroInQueue.pop()
            result.append(cur)
            for next in cur.nexts:
                inMap[next] = inMap[next] - 1
                if inMap[next] == 0:
                    zeroInQueue.add(next)
        return result
    
    def printSortedResult(self, result):
        if len(result) == 0:
            return
        for x in result:
            print(f"{x.value} ", end="")
        print()

def test():
    stack = Stack()
    stack.push(1)
    stack.push(2)
    stack.push(3)
    print(stack.pop())
    print(stack.pop())
    print(stack.pop())
    print(stack.pop())


def main():
    filePath = "/home/ubuntu/htl/Algorithm/python/class06/matrix.txt"
    graphGenerator = GraphGenerator()
    matrix = graphGenerator.createMatrix(filePath)
    graphGenerator.printMatrix(matrix)
    print("="*20)
    graph = graphGenerator.createGraph(matrix)
    graphGenerator.bfs(graph.nodes[1])
    graphGenerator.dfs(graph.nodes[1])
    print("="*20)
    graphGenerator.printSortedResult(graphGenerator.sortedTopolgy(graph))



main()